plugins/dell-kestrel/fu-dell-kestrel-ec.c

```diff
@@ -0,0 +1,1052 @@
+		g_prefix_error(error, "Failed to query dock info: ");
+		return FALSE;
+	}
+	if (res->len != length) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "invalid dock info size: expected %" G_GSIZE_FORMAT ",got %u",
+			    length,
+			    res->len);
+		return FALSE;
+	}
+
+	if (!fu_memcpy_safe((guint8 *)self->dock_info,
+			    length,
+			    0,
+			    res->data,
+			    res->len,
+			    0,
+			    length,
+			    error))
+		return FALSE;
+
+	if (!fu_dell_kestrel_ec_dock_info_extract(device, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_dell_kestrel_ec_dock_data_extract(FuDevice *device, GError **error)
+{
+	FuDellKestrelEc *self = FU_DELL_KESTREL_EC(device);
+	g_autofree gchar *mkt_name = NULL;
+	g_autofree gchar *serial = NULL;
+
+	/* set FuDevice name */
+	mkt_name = fu_struct_dell_kestrel_dock_data_get_marketing_name(self->dock_data);
+	fu_device_set_name(device, mkt_name);
+
+	/* set FuDevice serial */
+	serial =
+	    g_strdup_printf("%.7s/%016" G_GUINT64_FORMAT,
+			    fu_struct_dell_kestrel_dock_data_get_service_tag(self->dock_data),
+			    fu_struct_dell_kestrel_dock_data_get_module_serial(self->dock_data));
+	fu_device_set_serial(device, serial);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_kestrel_ec_dock_data_cmd(FuDevice *device, GError **error)
+{
+	FuDellKestrelEc *self = FU_DELL_KESTREL_EC(device);
+	FuDellKestrelEcHidCmd cmd = FU_DELL_KESTREL_EC_HID_CMD_GET_DOCK_DATA;
+	g_autoptr(GByteArray) res = fu_struct_dell_kestrel_dock_data_new();
+
+	/* get dock data over HID */
+	if (!fu_dell_kestrel_ec_read(device, cmd, res, error)) {
+		g_prefix_error(error, "Failed to query dock data: ");
+		return FALSE;
+	}
+
+	self->dock_data = fu_struct_dell_kestrel_dock_data_parse(res->data, res->len, 0, error);
+	if (!fu_dell_kestrel_ec_dock_data_extract(device, error))
+		return FALSE;
+	return TRUE;
+}
+
+gboolean
+fu_dell_kestrel_ec_is_dock_ready4update(FuDevice *device, GError **error)
+{
+	FuDellKestrelEc *self = FU_DELL_KESTREL_EC(device);
+	guint16 bitmask_fw_update_pending = 1 << 8;
+	guint32 dock_status = 0;
+
+	if (!fu_dell_kestrel_ec_dock_data_cmd(device, error))
+		return FALSE;
+
+	dock_status = fu_struct_dell_kestrel_dock_data_get_dock_status(self->dock_data);
+	if ((dock_status & bitmask_fw_update_pending) != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_BUSY,
+			    "dock status (%x) has pending updates, unavailable for now.",
+			    dock_status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dell_kestrel_ec_own_dock(FuDevice *device, gboolean lock, GError **error)
+{
+	FuDellKestrelEc *self = FU_DELL_KESTREL_EC(device);
+	g_autoptr(GByteArray) req = g_byte_array_new();
+	g_autoptr(GError) error_local = NULL;
+
+	fu_byte_array_append_uint8(req, FU_DELL_KESTREL_EC_HID_CMD_SET_MODIFY_LOCK);
+	fu_byte_array_append_uint8(req, 2); // length of data
+	fu_byte_array_append_uint16(req, lock ? 0xFFFF : 0x0000, G_LITTLE_ENDIAN);
```

Comment:
Rather than needing to "manually build" the command, it would be a lot cleaner and less error prone to have an rs struct for each command.  You can look at asus-hid for examples (see c0 c1 d0 commands)

Reply:
but I think this is how it was previously, this can be a follow up cleanup after merging, just keep in mind.